src/storage/lsm.rs

```diff
@@ -119,65 +127,107 @@ impl SSTable {
+        println!("New head commit log path is {:?}", commit_log_path);
         let commit_log = OpenOptions::new()
             .create(true)
             .write(true)
             .append(true)
-            .open(COMMIT_LOG_PATH)?;
+            .open(&commit_log_path)?;
 
-        let sstables: Vec<SSTable> = std::fs::read_dir(SSTABLE_DIR_PATH)?
-            .map(|res| res.map(|e| e.path()))
+        // Assume alphabetical order.
+        let sstables: Vec<SSTable> = std::fs::read_dir(SSTABLES_DIR_PATH)?
+            .map(|dir_iter| dir_iter.map(|e| e.path()))
             .map(|path| SSTable::read_from_file(path.unwrap()).unwrap())
             .collect();
 
         let ret = State {
             memtable,
-            commit_log: Some(commit_log),
+            commit_log_path,
+            commit_log,
+            memtable_in_flush: None,
             sstables,
         };
         Ok(ret)
     }
 
-    pub fn flush_memtable(&mut self) -> Result<()> {
-        let path = PathBuf::from(format!(
-            "/tmp/pancake/sstables/{}.data",
-            SystemTime::now()
-                .duration_since(UNIX_EPOCH)
-                .unwrap()
-                .as_micros()
-        ));
-        let sstable = SSTable::write_from_memtable(&self.memtable, path)?;
-        self.sstables.push(sstable);
-        self.memtable.0.clear();
-        self.commit_log.take(); // Close the file.
-        self.commit_log = Some(
-            OpenOptions::new()
-                .create(true)
-                .write(true)
-                .append(true)
-                .open(COMMIT_LOG_PATH)?,
-        );
+    fn flush_memtable(&mut self) -> Result<()> {
+        let new_cl_path = new_path(COMMIT_LOGS_DIR_PATH);
+        let new_cl = OpenOptions::new()
+            .create(true)
+            .write(true)
+            .append(true)
+            .open(&new_cl_path)?;
+        let old_cl_path: PathBuf;
+        {
+            // TODO MutexGuard here
+            let old_mt = mem::replace(&mut self.memtable, Memtable::default());
+            self.memtable_in_flush = Some(old_mt);
+
+            self.commit_log = new_cl;
+            old_cl_path = mem::replace(&mut self.commit_log_path, new_cl_path);
+        }
+
+        let new_sst = SSTable::write_from_memtable(self.memtable_in_flush.as_ref().unwrap(), new_path(SSTABLES_DIR_PATH))?;
+        {
+            // TODO MutexGuard here
+            self.sstables.push(new_sst);
+            self.memtable_in_flush.take();
+        }
+        println!("Removing {:?}", &old_cl_path);
+        fs::remove_file(old_cl_path)?;
 
         if self.sstables.len() >= SSTABLE_COMPACT_COUNT_THRESH {
-            // TODO compact
+            self.compact_sstables()?;
+        }
+
+        Ok(())
+    }
+
+    fn compact_sstables(&mut self) -> Result<()> {
+        let mut dense_idx = Memtable::default();
+        for old_sst in self.sstables.iter() {
```

Comment:
Good catch! In the current form, I'm conflating lack of value with tombstone value. This is fixed in https://github.com/ysono/pancake/pull/8/.

And yes the intention is to read sstables in the order of old->new, and the same key may exist in multiple sstables, but newer info overrides old info. Caveat is that we're assuming that a directory scan returns files in the alphabetical order.

Reply:
ok. i just checked the contract on `read_dir`.

> The order in which this iterator returns entries is platform and filesystem dependent.

https://doc.rust-lang.org/std/fs/fn.read_dir.html#platform-specific-behavior

-----

so it will be necessary to enforce the desired property by sorting after this `map` invocation here:
https://github.com/ysono/pancake/blob/1dc4dea949aba3d5ff3d2f8ac8ff175af82c1d03/src/storage/lsm.rs#L161

and also here:
https://github.com/ysono/pancake/blob/1dc4dea949aba3d5ff3d2f8ac8ff175af82c1d03/src/storage/lsm.rs#L143

Reply:
Thanks for digging in. Another item for a fs util mod.